ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topk/TopkKudaf.java

```diff
@@ -61,49 +76,38 @@
       return currentAggVal;
     }
 
-    int nullIndex = ArrayUtil.getNullIndex(currentAggVal);
+    final int nullIndex = ArrayUtil.getNullIndex(currentAggVal);
     if (nullIndex != -1) {
       currentAggVal[nullIndex] = currentVal;
-      Arrays.sort(currentAggVal, comparator());
+      Arrays.sort(currentAggVal, comparator);
       return currentAggVal;
     }
-    System.arraycopy(currentAggVal, 0, tempTopKArray, 0, topKSize);
-    tempTopKArray[topKSize] = currentVal;
-    Arrays.sort(tempTopKArray, Collections.reverseOrder());
-    return Arrays.copyOf((T[]) tempTopKArray, topKSize);
-  }
 
-  public static <T> Comparator<T> comparator() {
-    return (v1, v2) -> {
-      if(v1 == null && v2 == null) {
-        return 0;
-      }
-      if (v1 == null) {
-        return 1;
-      }
-      if (v2 == null) {
-        return -1;
-      }
+    final T last = currentAggVal[currentAggVal.length - 1];
+    if (currentVal.compareTo(last) <= 0) {
+      return currentAggVal;
+    }
```

Comment:
Good optimization!

Reply:
Yer, makes a _lot_ of sense, right?  :D

Reply:
It also means we can do away with the temp buffer that was causing the class to not be thread safe.